From b760a121d4826671b60e96b7c01906e45c964237 Mon Sep 17 00:00:00 2001
From: Pankaj Dev <pankaj.dev@st.com>
Date: Sat, 30 Dec 2023 11:06:52 +0530
Subject: [PATCH 526/954] feat(st-usb): enable USBOTG DFU for STM32MP21

USBOTG DFU mode is present on STM32MP21. Adapt device start and stop
sequences using STM32MP21 flag.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 00c51c8279e33ac2e5ce0dd68df68696c10b33c5 feat(st-usb): enable USBOTG DFU for STM32MP21]
Signed-off-by: Pankaj Dev <pankaj.dev@st.com>
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Change-Id: I6bdf516f737b59b197f54f6d4f55f643f51dba8b
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/361430
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/st/usb/stm32mp1_usb.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/st/usb/stm32mp1_usb.c b/drivers/st/usb/stm32mp1_usb.c
index 78890f5eb2..2b792c1524 100644
--- a/drivers/st/usb/stm32mp1_usb.c
+++ b/drivers/st/usb/stm32mp1_usb.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, STMicroelectronics - All Rights Reserved
+ * Copyright (c) 2021-2024, STMicroelectronics - All Rights Reserved
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -35,6 +35,7 @@
 #define OTG_GINTSTS				0x014U
 #define OTG_GINTMSK				0x018U
 #define OTG_GRXSTSP				0x020U
+#define OTG_GGPIO				0x038U
 #define OTG_GLPMCFG				0x054U
 #define OTG_DCFG				0x800U
 #define OTG_DCTL				0x804U
@@ -124,6 +125,9 @@
 #define STS_SETUP_COMP				4U
 #define STS_SETUP_UPDT				6U
 
+/* Bit definitions for OTG_GGPIO register */
+#define OTG_GGPIO_IDPULLUP_DIS			BIT(28)
+
 /* Bit definitions for OTG_GLPMCFG register */
 #define OTG_GLPMCFG_BESL			GENMASK(5, 2)
 
@@ -700,6 +704,10 @@ static enum usb_status usb_dwc2_stop_device(void *handle)
 	/* Disconnect the USB device by disabling the pull-up/pull-down */
 	mmio_setbits_32((uintptr_t)handle + OTG_DCTL, OTG_DCTL_SDIS);
 
+#if STM32MP21
+	mmio_clrbits_32(usb_base_addr + OTG_GGPIO, OTG_GGPIO_IDPULLUP_DIS);
+#endif
+
 	return USBD_OK;
 }
 
@@ -1055,7 +1063,9 @@ static enum usb_status usb_dwc2_start_device(void *handle)
 
 	mmio_clrbits_32(usb_base_addr + OTG_DCTL, OTG_DCTL_SDIS);
 	mmio_setbits_32(usb_base_addr + OTG_GAHBCFG, OTG_GAHBCFG_GINT);
-
+#if STM32MP21
+	mmio_setbits_32(usb_base_addr + OTG_GGPIO, OTG_GGPIO_IDPULLUP_DIS);
+#endif
 	return USBD_OK;
 }
 
-- 
2.53.0

